packages/flutter_tools/lib/src/base/multi_root_file_system.dart

```diff
@@ -0,0 +1,273 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io' as io
+    show
+        Directory,
+        File,
+        FileStat,
+        FileSystemEntity,
+        FileSystemEntityType,
+        Link;
+
+import 'package:file/file.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as p; // flutter_ignore: package_path_import
+
+/// A [FileSystem] that wraps the [delegate] file system to create an overlay of
+/// files from multiple [roots].
+///
+/// Regular paths or `file:` URIs are resolved directly in the underlying file
+/// system, but URIs that use a special [scheme] are resolved by searching
+/// under a set of given roots in order.
+///
+/// For example, consider the following inputs:
+///
+///   - scheme is `multi-root`
+///   - the set of roots are `/a` and `/b`
+///   - the underlying file system contains files:
+///         /root_a/dir/only_a.dart
+///         /root_a/dir/both.dart
+///         /root_b/dir/only_b.dart
+///         /root_b/dir/both.dart
+///         /other/other.dart
+///
+/// Then:
+///
+///   - file:///other/other.dart is resolved as /other/other.dart
+///   - multi-root:///dir/only_a.dart is resolved as /root_a/dir/only_a.dart
+///   - multi-root:///dir/only_b.dart is resolved as /root_b/dir/only_b.dart
+///   - multi-root:///dir/both.dart is resolved as /root_a/dir/only_a.dart
+class MultiRootFileSystem extends ForwardingFileSystem {
+  MultiRootFileSystem({
+    required FileSystem delegate,
+    required String scheme,
+    required List<String> roots,
+  })   : assert(delegate != null),
+        assert(roots.isNotEmpty),
+        _scheme = scheme,
+        _roots =
+            roots.map((String root) => delegate.path.normalize(root)).toList(),
+        super(delegate);
+
+  @visibleForTesting
+  FileSystem get fileSystem => delegate;
+
+  final String _scheme;
+  final List<String> _roots;
+
+  @override
+  File file(dynamic path) => MultiRootFile(
```

Comment:
here and below I would try for a 2 space indent

Reply:
Updated to match Flutter code style.